File: common/models/monitor.go
/*
 * Copyright (c) 2019 Huawei Technologies Co., Ltd.
 * A-Tune is licensed under the Mulan PSL v2.
 * You can use this software according to the terms and conditions of the Mulan PSL v2.
 * You may obtain a copy of Mulan PSL v2 at:
 *     http://license.coscl.org.cn/MulanPSL2
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY OR FIT FOR A PARTICULAR
 * PURPOSE.
 * See the Mulan PSL v2 for more details.
 * Create: 2019-10-29
 */

package models

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"strings"
	
	"gitee.com/openeuler/A-Tune/common/config"
	HTTP "gitee.com/openeuler/A-Tune/common/http"
	"gitee.com/openeuler/A-Tune/common/log"
)

// MonitorBody :the body send to monitor service
type MonitorBody struct {
	Module  string `json:"module"`
	Purpose string `json:"purpose"`
	Fmt     string `json:"fmt"`
	Path    string `json:"path"`
	Para    string `json:"para"`
}

// MonitorGetResp :the response of the monitor service
type MonitorGetResp struct {
	Status string `json:"status"`
	Value  string `json:"value"`
}

// Get method calling the get method of the monitor service
func (m *MonitorBody) Get() (*MonitorGetResp, error) {
	url := config.GetURL(config.MonitorURI)
	res, err := HTTP.Get(url, m)
	if err != nil {
		return nil, err
	}

	defer res.Body.Close()
	if res.StatusCode != 200 {
		return nil, fmt.Errorf("connect to monitor service failed")
	}

	resBody, err := ioutil.ReadAll(res.Body)
	if err != nil {
		return nil, err
	}

	respPutIns := new(MonitorGetResp)
	err = json.Unmarshal(resBody, respPutIns)
	if err != nil {
		return nil, err
	}

	return respPutIns, nil
}

// MonitorGet method calling the get method of the monitor service, return error if failed
func MonitorGet(module string, purpose string, format string, path string, para string) (string, error) {
	monitorBody := &MonitorBody{
		Module:  strings.ToUpper(module),
		Purpose: strings.ToUpper(purpose),
		Fmt:     format,
		Path:    path,
		Para:    para,
	}

	respGetIns, err := monitorBody.Get()
	if err != nil {
		log.Errorf("model monitor module %s get data failed: %v", module, err)
		return "", err
	}

	if respGetIns.Status != "OK" {
		log.Errorf("model monitor module %s get data failed, error: %s", module, respGetIns.Value)
		return "", fmt.Errorf(respGetIns.Value)
	}
	return respGetIns.Value, nil
}

File: analysis/atuned/monitor.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-
# Copyright (c) 2019 Huawei Technologies Co., Ltd.
# A-Tune is licensed under the Mulan PSL v2.
# You can use this software according to the terms and conditions of the Mulan PSL v2.
# You may obtain a copy of Mulan PSL v2 at:
#     http://license.coscl.org.cn/MulanPSL2
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY OR FIT FOR A PARTICULAR
# PURPOSE.
# See the Mulan PSL v2 for more details.
# Create: 2019-10-29

"""
Restful api for monitor, in order to provide the method of post and get.
"""

import getopt
from flask import abort
from flask import current_app
from flask_restful import reqparse, Resource

from analysis.atuned import MPI_INSTANCE
from analysis.atuned.parser import MONITOR_GET_PARSER
from analysis.atuned.parser import MONITOR_POST_PARSER

PARSER = reqparse.RequestParser()


class Monitor(Resource):
    """restful api for monitor, in order to provide the method of post and get"""
    module = "module"
    purpose = "purpose"

    def get(self):
        """provide the method of get"""
        result = {}
        args = MONITOR_GET_PARSER.parse_args()
        current_app.logger.info(args)
        module = args.get(self.module)
        purpose = args.get(self.purpose, None)
        fmt = args.get("fmt")
        path = args.get("path")
        para = args.get("para")

        path = None if path.strip() == "" else path
        para = None if para.strip() == "" else para
        monitors = MPI_INSTANCE.get_monitors(module, purpose)
        if len(monitors) < 1:
            result["status"] = f"module: {module}, purpose:{purpose} is not exist"
            return result, 200

        monitor = monitors[0]
        ret = monitor.report(fmt, path, para)
        if isinstance(ret, Exception):
            result["value"] = str(ret)
            result["status"] = "FAILED"
            return result, 200

        result["status"] = "OK"
        result["value"] = ret
        return result, 200

    def post(self):
        """provide the method of post"""
        result = {}
        args = MONITOR_POST_PARSER.parse_args()
        current_app.logger.info(args)

        monitors = MPI_INSTANCE.get_monitors(args.get(self.module), args.get(self.purpose, None))

        if len(monitors) < 1:
            abort(404)

        monitor = monitors[0]
        ret = monitor.report("data", None, args.get("field"))
        opts, args = getopt.getopt(args.get("field").split(";")[1].split(), None,
                                   ['fields=', 'addr-merge='])
        opt = []
        for option, value in opts:
            if option in "--fields":
                opt.append(value)

        events = []
        for index, _ in enumerate(ret):
            if isinstance(ret[index], list):
                event = {}
                for i in range(len(ret[index])):
                    event[opt[i]] = ret[index][i]
                events.append(event)
            else:
                result[opt[index]] = ret[index]

        if len(events) > 0:
            result["data"] = events
        return result, 200
